### PR TITLE
Support variables and classes with meta information in CMDB data

### PR DESCRIPTION
**Pull request by vpodzime**

### CODE REVIEW

**Comment by vpodzime** [comment]
@cf-bottom jenkins with exotics, please

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/6538/)

(with exotics)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/6538/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-6538/

File: libpromises/eval_context.c [hunk 1]
```diff
@@ -1751,6 +1765,15 @@ bool EvalContextClassPutSoftNS(EvalContext *ctx, const char *ns, const char *nam
     return EvalContextClassPut(ctx, ns, name, true, scope, tags);
 }
 
+/**
+ * Takes over #tags in case of success.
```
Comment (larsewi) [inline libpromises/eval_context.c:1780]:
```C
/**
 * Takes ownership of #tags in case of success.
 */
```

File: libpromises/cmdb.c [hunk 2]
```diff
@@ -194,6 +201,93 @@ static bool ReadCMDBVars(EvalContext *ctx, JsonElement *vars)
...
+{
+    StringSet *tags = NULL;
+    if (json_tags != NULL)
+    {
+        if ((JsonGetType(json_tags) != JSON_TYPE_ARRAY) ||
+            (!JsonArrayContainsOnlyPrimitives((JsonElement*) json_tags)))
+        {
+            Log(LOG_LEVEL_ERR,
+                "Invalid json_tags information for %s '%s' in CMDB data:"
+                " must be a JSON array of strings",
```
Comment (larsewi) [inline libpromises/cmdb.c:217]: You don't check whether or not the primitives are strings?

Comment (vpodzime) [reply libpromises/cmdb.c:217]: Not sure if I need to care, all primitives are strings internally and all can/will be converted to strings by `JsonArrayToStringSet()` below. If user wants numbers as tags, they can do so.

File: libpromises/eval_context.c (hunk 1 above)
Comment (vpodzime) [reply libpromises/eval_context.c:1780]: I don't see the difference between the two sentences.

**Comment by vpodzime** [comment]
jenkins without exotics: [Build Status](https://ci.cfengine.com/job/pr-pipeline/6547/)

File: libpromises/generic_agent.c [hunk 3]
```diff
@@ -348,6 +392,130 @@ static bool LoadAugmentsData(EvalContext *ctx, const char *filename, const JsonE
...
+
+            if (variables == NULL || JsonGetType(variables) != JSON_TYPE_OBJECT)
+            {
+                Log(LOG_LEVEL_ERR, "Invalid augments variables in '%s', must be a JSON object", filename);
+                goto variables_cleanup;
+            }
+
+            JsonIterator variables_iter = JsonIteratorInit(variables);
+            const char *vkey;
+            while ((vkey = JsonIteratorNextKey(&variables_iter)))
```
Comment (larsewi) [inline libpromises/generic_agent.c:409]: Don't include assignments inside expressions, unless they avoid great repetition & use explicit comparisons.
 - [CONTIBUTING.md](https://github.com/cfengine/core/blob/master/CONTRIBUTING.md)

File: libpromises/cmdb.c (hunk 2 above)
Comment (larsewi) [reply libpromises/cmdb.c:217]: Okey, thanks for clarification :smile:

File: libpromises/eval_context.c (hunk 1 above)
Comment (larsewi) [reply libpromises/eval_context.c:1780]: The latter sounded more precise in my head. So, I just thought I'd suggest it :wink:

File: libpromises/generic_agent.c (hunk 3 above)
Comment (vpodzime) [reply libpromises/generic_agent.c:409]: This is just copy-pasted from the nearby code, I'm not changing it there. Besides, I think iterations like this should be an exception.

File: libpromises/cmdb.c (hunk 2 above)
Comment (larsewi) [reply libpromises/cmdb.c:217]: Should we maybe include an edge-case in the acceptance test where one of these other primitives are used?

Comment (vpodzime) [reply libpromises/cmdb.c:217]: I don't think that's necessary. We have a test for `JsonArrayToStringSet()` with a non-string primitive.

Comment (larsewi) [reply libpromises/cmdb.c:217]: Ah, okey. Nice!

**Review by larsewi — APPROVED** [review]
ACK

**Comment by olehermanse** [comment]
@vpodzime merging your other PR created a conflict: https://github.com/cfengine/core/pull/4625

**Review by olehermanse — APPROVED** [review]
LGTM! 🚀
